Fix `DOMFilterFactory.#createUrl` in MOZCENTRAL builds (18417 PR follow-up)

Snuffleupagus opened the pull request:
Somehow I managed to mess up the URL creation relevant to e.g. MOZCENTRAL builds, which is breaking the pending PDF.js update in mozilla-central; sorry about that!

Snuffleupagus committed:
Fix `DOMFilterFactory.#createUrl` in MOZCENTRAL builds (18417 PR follow-up)

Somehow I managed to mess up the URL creation relevant to e.g. MOZCENTRAL builds, which is breaking the pending PDF.js update in mozilla-central; sorry about that!

To avoid future issues, we'll now always check if absolute filter-URLs are necessary regardless of the build-target.

timvandermeij reviewed (approved):
r=me, with passing tests.

timvandermeij commented:
/botio test

calixteman reviewed (approved):
Arrff... I missed that too.
Thank you.

calixteman commented:
This test should have failed: https://github.com/mozilla/pdf.js/blob/2d254375a8ca8babf8a63685b39bd09388de9f80/test/test_manifest.json#L8777-L8782.
@Snuffleupagus do you know why it didn't ?

Snuffleupagus commented:
> do you know why it didn't ?

Yes, because of the pre-processor; see https://github.com/mozilla/pdf.js/blob/2d254375a8ca8babf8a63685b39bd09388de9f80/src/display/display_utils.js#L126-L143

The idea was to ensure that we actually test the GENERIC code in the PDF.js test-suite, since the Firefox code-path is covered by tests in mozilla-central. Perhaps we should just use the `#baseUrl` feature regardless of build-target here?

Snuffleupagus commented:
@calixteman Would you prefer the following diff instead, where we'd *always* check if a `baseUrl` is needed?
```diff
diff --git a/src/display/display_utils.js b/src/display/display_utils.js
index 30964c9ab..9d4a41723 100644
--- a/src/display/display_utils.js
+++ b/src/display/display_utils.js
@@ -124,22 +124,19 @@ class DOMFilterFactory extends BaseFilterFactory {
   }
 
   #createUrl(id) {
-    if (typeof PDFJSDev === "undefined" || PDFJSDev.test("GENERIC")) {
-      if (this.#baseUrl === undefined) {
-        const url = this.#document.URL;
-        if (url === this.#document.baseURI) {
-          // No `<base>`-element present, hence a relative URL should work.
-          this.#baseUrl = "";
-        } else if (isDataScheme(url)) {
-          warn('#createUrl: ignore "data:"-URL for performance reasons.');
-          this.#baseUrl = "";
-        } else {
-          this.#baseUrl = url.split("#", 1)[0];
-        }
+    if (this.#baseUrl === undefined) {
+      const url = this.#document.URL;
+      if (url === this.#document.baseURI) {
+        // No `<base>`-element present, hence a relative URL should work.
+        this.#baseUrl = "";
+      } else if (isDataScheme(url)) {
+        warn('#createUrl: ignore "data:"-URL for performance reasons.');
+        this.#baseUrl = "";
+      } else {
+        this.#baseUrl = url.split("#", 1)[0];
       }
-      return `url(${this.#baseUrl}#${id})`;
     }
-    return `url(${id})`;
+    return `url(${this.#baseUrl}#${id})`;
   }
 
   addFilter(maps) {
```

calixteman commented:
> @calixteman Would you prefer the following diff instead, where we'd _always_ check if a `baseUrl` is needed?
> 
> ```diff
> diff --git a/src/display/display_utils.js b/src/display/display_utils.js
> index 30964c9ab..9d4a41723 100644
> --- a/src/display/display_utils.js
> +++ b/src/display/display_utils.js
> @@ -124,22 +124,19 @@ class DOMFilterFactory extends BaseFilterFactory {
>    }
>  
>    #createUrl(id) {
> -    if (typeof PDFJSDev === "undefined" || PDFJSDev.test("GENERIC")) {
> -      if (this.#baseUrl === undefined) {
> -        const url = this.#document.URL;
> -        if (url === this.#document.baseURI) {
> -          // No `<base>`-element present, hence a relative URL should work.
> -          this.#baseUrl = "";
> -        } else if (isDataScheme(url)) {
> -          warn('#createUrl: ignore "data:"-URL for performance reasons.');
> -          this.#baseUrl = "";
> -        } else {
> -          this.#baseUrl = url.split("#", 1)[0];
> -        }
> +    if (this.#baseUrl === undefined) {
> +      const url = this.#document.URL;
> +      if (url === this.#document.baseURI) {
> +        // No `<base>`-element present, hence a relative URL should work.
> +        this.#baseUrl = "";
> +      } else if (isDataScheme(url)) {
> +        warn('#createUrl: ignore "data:"-URL for performance reasons.');
> +        this.#baseUrl = "";
> +      } else {
> +        this.#baseUrl = url.split("#", 1)[0];
>        }
> -      return `url(${this.#baseUrl}#${id})`;
>      }
> -    return `url(${id})`;
> +    return `url(${this.#baseUrl}#${id})`;
>    }
>  
>    addFilter(maps) {
> ```

Let's do that: I think it's slightly simpler and it shouldn't add any specific penalty.

calixteman commented:
@Snuffleupagus, could you merge this PR asap, and then I'll make a new release.

Snuffleupagus commented:
/botio test
